File: lab2/max.cpp
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

int findMax(int* array, int left, int right) {
    if (left == right) {
        return array[left];
    }
    
    int mid = left + (right - left) / 2;

    int leftMax = findMax(array, left, mid);
    int rightMax = findMax(array, mid + 1, right);

    return max(leftMax, rightMax);
}

void print(int* array, int n) {
    for (int i = 0; i < n; i++) {
        cout << array[i] << " ";
    }
    cout << endl;
}

void testMaxFromFile(const string& filename) {
    ifstream infile(filename);

    if (!infile.is_open()) {
        cout << "Error opening file: " << filename << endl;
        return;
    }

    string line;
    while (getline(infile, line)) {
        istringstream iss(line);
        int count = 0;

        int num;
        while (iss >> num) {
            count++;
        }

        iss.clear();
        iss.seekg(0, ios::beg);

        int* array = new int[count];
        for (int i = 0; i < count; i++) {
            iss >> array[i];
        }

        if (count > 0) {
            cout << "Original array: ";
            print(array, count);

            int max_value = findMax(array, 0, count - 1);
            cout << "Maximum value: " << max_value << endl;
            cout << endl;
        }

        delete[] array;  
    }

    infile.close();
}

int main() {
    testMaxFromFile("sorttest.txt");
    return 0;
}

File: lab2/max.py
def find_max(array, left, right):
    if left == right:
        return array[left]

    mid = (left + right) // 2

    left_max = find_max(array, left, mid)
    right_max = find_max(array, mid + 1, right)

    return max(left_max, right_max)

def test_max_from_file(filename):
    try:
        with open(filename, 'r') as file:
            for line in file:
                array = list(map(int, line.split()))

                if array:
                    max_value = find_max(array, 0, len(array) - 1)
                    print("Original array:", ' '.join(map(str, array)))
                    print("Maximum value:", max_value)
                    print()
    except FileNotFoundError:
        print(f"Error opening file: {filename}")

if __name__ == "__main__":
    test_max_from_file("sorttest.txt")
